src/Features/Core/Portable/Ambiguity/AbstractAmbiguousTypeCodeFixProvider.cs

```diff
@@ -0,0 +1,103 @@
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.AmbiguityCodeFixProvider
+{
+    internal abstract class AbstractAmbiguousTypeCodeFixProvider : CodeFixProvider
+    {
+        protected abstract SyntaxNode GetAliasDirective(string typeName, ISymbol symbol);
+
+        public override FixAllProvider GetFixAllProvider() => null;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Innermost: We are looking for an IdentifierName. IdentifierName is sometimes at the same span as its parent (e.g. SimpleBaseTypeSyntax).
+            var diagnosticNode = root.FindNode(context.Span, getInnermostNodeForTie: true);
+            if (!syntaxFacts.IsIdentifierName(diagnosticNode))
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var symbolInfo = semanticModel.GetSymbolInfo(diagnosticNode, cancellationToken);
+            if (SymbolCandidatesContainsSupportedSymbols(symbolInfo))
+            {
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var diagnostic = context.Diagnostics.First();
+                var compilation = semanticModel.Compilation;
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+                var typeName = GetAliasFromDiagnosticNode(syntaxFacts, diagnosticNode);
+                foreach (var symbol in symbolInfo.CandidateSymbols)
+                {
+                    var aliasDirective = GetAliasDirective(typeName, symbol);
+                    var codeActionPreviewText = await GetTextPreviewOfChangeAsync(aliasDirective,
+                                                                                  document.Project.Solution.Workspace,
+                                                                                  optionSet,
+                                                                                  cancellationToken).ConfigureAwait(false);
+                    var newRoot = addImportService.AddImport(compilation, root, diagnosticNode, aliasDirective, placeSystemNamespaceFirst);
```

Comment:
nit, move this code into the lambda below to make RegisterCodeFixAsync do as little work as possible.